src/fileformat/pfsinppm.1
.TH "pfsinppm" 1
.SH NAME
pfsinppm \- Load images or frames in PBM formats
.SH SYNOPSIS
.B pfsinppm
(<file> [--linear] [--absolute <max_lum>] [--frames <range>] [--skip-missing])  [<file>...]

.SH DESCRIPTION
.I pfsinppm
command loads images in PBM formats (PPM, PNM or PGM) and writes
\fIpfs\fR stream to the standard output. The \fIpfs\fR stream is
usually piped to another program for further processing. To detect the
format automatically based on the extension, use \fIpfsin\fR
command. For more information on PBM formats, refer to the NetPBM web page
(http://netpbm.sourceforge.net/).
.PP
Note that PPM or PNM images are low dynamic range. Therefore pixel
values (0-255) are scaled to 0-1 before
storing them in pfs stream. Similarly, before writing low dynamic
range image from pfs stream, pixel values are multiplied by 255.  By
default, the 'LUMINANCE' tag is set to 'DISPLAY'. The '--linear' switch can force the
inverse sRGB transformation and provide linear data. In this case the 'LUMINANCE'
tag is set to 'RELATIVE'. '--absolute' switch can
be used to convert pixels to absolute luminance values.
.PP
To read images from standard input use a single dash '-' instead of
filename. The images are read until EOF is reached.
.PP
Each file can contain a \%%d pattern, which is substituted with frame
numbers. The pattern has the same syntax as C
.I printf
command. For example, you can use \%%04d to make the frame number
four digit with proceedings zeros. See the OPTIONS section below for details.
.SH OPTIONS
.TP
.B \--frames <range>
Range is given in mathlab / octave format:

.B "startframe:step:endframe"

Frame numbers start with
.B "startframe"
(default 0), are increased by
.B "step"
(default 1) and stop at
.B "endframe"
You can skip one of those values, for example
.I "1:100"
for frames 1,2,...,100 and
.I 0:2:
for frame 0,2,4,... up to the last file that exists.

.TP
.B \--skip-missing
Skip up to ten frames in a row if corresponding files are
missing. Otherwise the program stops reading sequence at the first
file that does not exists. This switch does not apply to the first
frame in a sequence. This switch can be useful if there is a rendered
animation where some of the frame has not been generated.

.TP
.B \--linear, -l
Converts pixel values to linear luminance (XYZ), assuming the sRGB
color space for the input image. The maximum pixel value (255,255,255)
is mapped to Y=1. \fILUMINANCE\fR tag is set to RELATIVE.

.TP
.B \--absolute <max_lum>, -a <max_lum>
\fB--absolute\fR converts pixel values to an absolute linear luminance
(XYZ), that is the color space, in which channel Y contains luminance
given in cd/m^2. The sRGB color space is assumed for the input
image. The maximum pixel value (255,255,255) is mapped to
Y=\fI<max_lum>\fR. \fI<max_lum>\fR is typically set to 80 [cd/m^2] for
a CRT monitor. \fILUMINANCE\fR tag is set to
ABSOLUTE. \fB--absolute\fR process images almost the same as
\fB--relative\fR, but additionally it scales all pixels by
\fI<max_lum>\fR.

.SH EXAMPLES
.TP
pfsinppm frame\%%04d.ppm \--frames 0:10 | pfsview

Read frames from files frame0000.ppm, frame0001.ppm, ...,
frame0010.ppm and show them using pfsview.

.SH BUGS
Please report bugs and comments on implementation to 
the discussion group http://groups.google.com/group/pfstools
.SH "SEE ALSO"
.BR pfsin (1),
.BR pfsout (1)




NEWS
This file is not maintained, but required by automake. Visit pfstools
web page to find the latest news:

http://pfstools.sourceforge.net/



src/fileformat/pfsinppm.cpp
/**
 * @brief Read files in PPM format
 * 
 * This file is a part of PFSTOOLS package.
 * ---------------------------------------------------------------------- 
 * Copyright (C) 2003,2004 Rafal Mantiuk and Grzegorz Krawczyk
 * 
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 * ---------------------------------------------------------------------- 
 * 
 * @author Rafal Mantiuk, <mantiuk@mpi-sb.mpg.de>
 * @author Grzegorz Krawczyk, <krawczyk@mpi-sb.mpg.de>
 *
 * $Id: pfsinppm.cpp,v 1.6 2009/05/13 17:26:14 rafm Exp $
 */

#include <config.h>

#include <cstdlib>

#include <iostream>

#include <getopt.h>
#include <pfs.h>

#include "ppmio.h"

#define PROG_NAME "pfsinppm"


class QuietException 
{
};

void printHelp()
{
  std::cerr << PROG_NAME " [--linear] [--absolute <max_lum>] [--verbose] [--help]" << std::endl
            << "See man page for more information." << std::endl;
}



void readFrames( int argc, char* argv[] )
{
  pfs::DOMIO pfsio;

  bool verbose = false;
  bool opt_linear=false;
  float absoluteMaxLum = 0;
  
  // Parse command line parameters
  static struct option cmdLineOptions[] = {
    { "help", no_argument, NULL, 'h' },
    { "verbose", no_argument, NULL, 'v' },
    { "linear", no_argument, NULL, 'l' },
    { "absolute", required_argument, NULL, 'a' },
    { NULL, 0, NULL, 0 }
  };
  static const char optstring[] = "lhva:";
    
  pfs::FrameFileIterator it( argc, argv, "rb", NULL, stdin,
    optstring, cmdLineOptions );
    
  int optionIndex = 0;
  while( 1 ) {
    int c = getopt_long (argc, argv, optstring, cmdLineOptions, &optionIndex);
    if( c == -1 ) break;
    switch( c ) {
    case 'h':
      printHelp();
      throw QuietException();
    case 'v':
      verbose = true;
      break;
    case 'l':
      opt_linear = true;
      break;
    case 'a':
      absoluteMaxLum = (float)strtod( optarg, NULL );
      break;
    case '?':
      throw QuietException();
    case ':':
      throw QuietException();
    }
  }

  if( absoluteMaxLum != 0 && opt_linear )
    throw pfs::Exception( "'absolute' and 'linear' are conflicting options" );
  
  if( absoluteMaxLum < 0 )
    throw pfs::Exception( "maximum absolute luminance must be > 0" );

  VERBOSE_STR << "linearize input image: " << ((opt_linear || absoluteMaxLum!=0) ? "yes" : "no") << std::endl;
  if( absoluteMaxLum != 0 )
    VERBOSE_STR << "maximum absolute luminance: " << absoluteMaxLum << std::endl;

  bool eofP = false;
  while( !eofP ) {
    pfs::FrameFile ff = it.getNextFrameFile();
    if( ff.fh == NULL ) break; // No more frames

    PPMReader reader( PROG_NAME, ff.fh );

    VERBOSE_STR << "reading file '" << ff.fileName << "'" << std::endl;

    pfs::Frame *frame = pfsio.createFrame( reader.getWidth(),
      reader.getHeight() );
    pfs::Channel *X, *Y, *Z;
    frame->createXYZChannels( X, Y, Z );

    //Store sRGB data temporarily in XYZ channels        
    eofP = reader.readImage( X, Y, Z );
    if( opt_linear || absoluteMaxLum != 0 )
    {
      pfs::transformColorSpace( pfs::CS_SRGB, X, Y, Z, pfs::CS_XYZ, X, Y, Z );
      if( absoluteMaxLum != 0 ) {
        // Rescale to absolute luminance level
        const int pixCount = X->getWidth()*X->getHeight();
        for( int i = 0; i < pixCount; i++ ) {
          (*X)(i) *= absoluteMaxLum;
          (*Y)(i) *= absoluteMaxLum;
          (*Z)(i) *= absoluteMaxLum;
        }        
        frame->getTags()->setString("LUMINANCE", "ABSOLUTE");
      } else
        frame->getTags()->setString("LUMINANCE", "RELATIVE");
    }
    else
    {
      pfs::transformColorSpace( pfs::CS_RGB, X, Y, Z, pfs::CS_XYZ, X, Y, Z );
      frame->getTags()->setString("LUMINANCE", "DISPLAY");
    }

    // This is the luminance / luma perceived as reference white
    // Some tone-mappers may need this
    frame->getTags()->setString("WHITE_Y", "1");

    char strbuf[3];
    snprintf( strbuf, 3, "%d", reader.getBitDepth() );
    frame->getTags()->setString("BITDEPTH", strbuf );    

    const char *fileNameTag = strcmp( "-", ff.fileName )==0 ? "stdin" : ff.fileName;
    frame->getTags()->setString( "FILE_NAME", fileNameTag );
        
    pfsio.writeFrame( frame, stdout );
    pfsio.freeFrame( frame );
        
    it.closeFrameFile( ff );
  }
}


int main( int argc, char* argv[] )
{
  try {
    readFrames( argc, argv );
  }
  catch( pfs::Exception ex ) {
    std::cerr << PROG_NAME << " error: " << ex.getMessage() << std::endl;
    return EXIT_FAILURE;
  }        
  catch( QuietException  ex ) {
    return EXIT_FAILURE;
  }        
  return EXIT_SUCCESS;
}
